Firestore/core/test/firebase/firestore/remote/grpc_stream_test.cc

```diff
@@ -0,0 +1,249 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream.h"
+
+#include <initializer_list>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_completion.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/executor_std.h"
+#include "Firestore/core/test/firebase/firestore/util/grpc_tests_util.h"
+#include "absl/memory/memory.h"
+#include "grpcpp/client_context.h"
+#include "grpcpp/completion_queue.h"
+#include "grpcpp/create_channel.h"
+#include "grpcpp/generic/generic_stub.h"
+#include "grpcpp/support/byte_buffer.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using util::AsyncQueue;
+using util::GrpcStreamFixture;
+using util::OperationResult;
+using util::internal::ExecutorStd;
+using util::OperationResult::Error;
+using util::OperationResult::Ok;
+
+namespace {
+
+class Observer : public GrpcStreamObserver {
+ public:
+  void OnStreamStart() override {
+    observed_states.push_back("OnStreamStart");
+  }
+  void OnStreamRead(const grpc::ByteBuffer& message) override {
+    observed_states.push_back("OnStreamRead");
+  }
+  void OnStreamError(const util::Status& status) override {
+    observed_states.push_back("OnStreamError");
+  }
+
+  std::vector<std::string> observed_states;
+};
+
+}  // namespace
+
+class GrpcStreamTest : public testing::Test {
+ public:
+  GrpcStreamTest() : observer{absl::make_unique<Observer>()} {
+    fixture.CreateStream(observer.get());
+  }
+
+  ~GrpcStreamTest() {
+    fixture.Shutdown();
+  }
+
+  GrpcStream& stream() {
+    return fixture.stream();
+  }
+  AsyncQueue& async_queue() {
+    return fixture.async_queue();
+  }
+
+  void ForceFinish(std::initializer_list<OperationResult> results) {
+    fixture.ForceFinish(results);
+  }
+  void KeepPollingGrpcQueue() {
+    fixture.KeepPollingGrpcQueue();
+  }
+
+  const std::vector<std::string>& observed_states() const {
+    return observer->observed_states;
+  }
+
+  // This is to make `EXPECT_EQ` a little shorter and work around macro
+  // limitations related to initializer lists.
+  std::vector<std::string> States(std::initializer_list<std::string> states) {
+    return {states};
+  }
+
+  bool ObserverHas(const std::string& state) const {
+    return std::find(observed_states().begin(), observed_states().end(),
+                     state) != observed_states().end();
+  }
+
+  void StartStream() {
+    async_queue().EnqueueBlocking([&] { stream().Start(); });
+    ForceFinish({/*Start*/ Ok});
+  }
+
+ private:
+  GrpcStreamFixture fixture;
+  std::unique_ptr<Observer> observer;
+};
+
+TEST_F(GrpcStreamTest, CanFinishBeforeStarting) {
+  async_queue().EnqueueBlocking([&] { EXPECT_NO_THROW(stream().Finish()); });
```

Comment:
What prevents these tests from enqueueing all the actions here and immediately exiting before the other thread has a chance to execute any of the actions enqueued on the async queue?

More problematically, if there is a failure here, what prevents the error from being misattributed to the next case?

Shouldn't this be:

```c++
// In a test util header somewhere
template <typename T>
T Await(std::promise<T>* promise) {
  std::future<T> future = promise->get_future();
  std::future_status status = future.wait_for(kDefaultTestTimeout);
  if (status == std::future_status::ready) {
    return future.get();
  } else {
    EXPECT_EQ(status, std::future_status::ready) << "Test timed out after " << kDefaultTestTimeout;
    return {}  // Would require T to be default constructible; other options possible.
}

TEST_F(GrpcStreamTest, CanFinishBeforeStarting) {
  std::promise<void> finished;
  async_queue().EnqueueBlocking([&] {
    EXPECT_NO_THROW(stream().Finish());
    finished.set_value();
  });
  Await(&finished);
}
```

In XCTest we use XCTestExpectations liberally to do this kind of thing and I'd expect we'd need the same kind of thing here.